mantaray/backend.py

```diff
@@ -471,22 +487,22 @@ def _send_loop(self) -> None:
                     self._disconnect()  # stop recv loop
 
     def _connect(self) -> None:
-        assert self._sock is None
+        assert self._sock is None and not self._connected
 
         try:
             if self.ssl:
                 context = ssl.create_default_context()
-                sock: socket.socket | ssl.SSLSocket = context.wrap_socket(
+                self._sock = context.wrap_socket(
                     socket.socket(), server_hostname=self.host
                 )
             else:
-                sock = socket.socket()
-            sock.connect((self.host, self.port))
+                self._sock = socket.socket()
+            self._sock.connect((self.host, self.port))
+            self._connected = True
         except (OSError, ssl.SSLError) as e:
+            self._sock = None
```

Comment:
do not leak file descriptors
```suggestion
            if self._sock is not None:
                self._sock.close()
            self._sock = None
```